Load secrets from yml file

jonnyshaw89 opened the pull request:
### What is the context of this PR?
Load the application secrets from a .yml file on disk

### How to review 
Check that no secrets are defined as environment variables

adoublebarrel commented on app/__init__.py:
```diff
@@ -47,10 +48,16 @@ def __call__(self, environ, start_response):
+    secrets = yaml.safe_load(open(os.getenv('EQ_SECRETS')))
+    application.config.update(secrets)
+
+    if setting_overrides:
```
Default setting_overrides to a dict and you wont need the if...

adoublebarrel commented on app/__init__.py:
```diff
@@ -47,10 +48,16 @@ def __call__(self, environ, start_response):
     application = Flask(__name__, static_url_path='/s', static_folder='../static')
     application.config.from_object(settings)
 
+    secrets = yaml.safe_load(open(os.getenv('EQ_SECRETS')))
```
Would this be better off in application.py? Not sure create_app should have knowledge of where/how the secrets are kept.

jonnyshaw89 replied:
I initially had this in application.py as i thought it would then make testing easier but @ajmaddaford pointed out that it would be better if all of the loading of application config was in the same place.

adoublebarrel replied:
Yes that makes total sense. Longer term I was thinking that we'd move all config loading out of create_app to another place (e.g. application.py)? So create_app would only be concerned with putting the app together based off the config object|dict that's passed to it. That will help with [simplifying create_app](https://trello.com/c/90vVTmqt/425-look-at-reducing-the-complexity-of-create-app-app-init-py).

I'm probably jumping the gun a bit here though, happy with it as is.

adoublebarrel commented on tests/integration/create_token.py:
```diff
@@ -23,24 +22,28 @@
 
-    return generate_token(payload_vars)
+    def __init__(self, application):
+        self.application = application
```
Worth tightening this so that we just pass the private key, private key password and public key to the constructor? (looks like that's all it needs)

collisdigital reviewed (approved):
Looks good

adoublebarrel reviewed (commented):
Broadly looks good to me, just some minor suggestions.

adoublebarrel reviewed (approved):
LGTM

jonnyshaw89 commented:
No merge until https://github.com/ONSdigital/eq-terraform/pull/81 is in

ajmaddaford commented on app/secrets.py:
```diff
@@ -0,0 +1,24 @@
+def validate_required_secrets(secrets):
+    for required_secret in EXPECTED_SECRETS:
+        if required_secret not in secrets:
+            raise MissingSecretException("Missing Secret [{}]".format(required_secret))
```
I don't think we need to have a `MissingSecretException`, we could just use `Exception`. Also, the required_secret bit doesn't get printed, you just get `app.missing_secret_exception.MissingSecretException`. 

Replacing `MissingSecretException` with `Exception` works and prints something like `Exception: Missing Secret [EQ_SERVER_SIDE_STORAGE_DATABASE_URL]`

ajmaddaford commented on app/__init__.py:
```diff
@@ -50,10 +52,17 @@ def __call__(self, environ, start_response):
     application = Flask(__name__, static_url_path='/s', static_folder='../static')
     application.config.from_object(settings)
 
+    secrets = yaml.safe_load(open(os.getenv('EQ_SECRETS')))
```
I think we should rename this `EQ_SECRETS_FILE`

collisdigital commented:
Have you tried running this in Docker as well?

jonnyshaw89 commented:
@collisdigital yes I've tried it, but that doesn't even work on master.

jonnyshaw89 commented:
@collisdigital works after docker pull on base image

ajmaddaford commented on app/data_model/database.py:
```diff
@@ -68,15 +68,17 @@ def __repr__(self):
         self._db_session = self._create_session_and_engine_with_retry()
 
     def _create_session_and_engine(self):
+        logger.error("Database Address", database_url=self.database_url)
```
Debug logging?

ajmaddaford commented on docker-compose.yml:
```diff
@@ -31,9 +31,10 @@ services:
-      EQ_SERVER_SIDE_STORAGE_DATABASE_URL: postgres://digitaleq:digitaleq@db:5432/runner
-      EQ_RABBITMQ_URL: amqp://digitaleq:digitaleq@rabbit:5672/%2F
+      EQ_SECRETS_FILE: ".dev_secrets.yml"
+      EQ_SERVER_SIDE_STORAGE_DATABASE_URL: digitaleq:digitaleq@db:5432/runner
```
We should use the username and password environment variables here rather than overloading the URL

jonnyshaw89 replied:
They arn't settable via env var any more.

ajmaddaford commented on app/data_model/database.py:
```diff
@@ -68,15 +68,17 @@ def __repr__(self):
 class Database:
 
-    def __init__(self, database_url, database_setup_retry_count, database_setup_retry_delay):
+    def __init__(self, database_driver, database_url, username, password, database_setup_retry_count, database_setup_retry_delay):
```
We're calling two different things `database_url` here. We should rename `database_url` (AWS calls this an endpoint in RDS) or separate into host and port.

ajmaddaford commented on app/submitter/submitter.py:
```diff
@@ -16,9 +16,11 @@ def send_message(self, message, queue, tx_id):  # pylint: disable=unused-argumen
-        self.rabbitmq_url = rabbitmq_url
-        self.rabbitmq_secondary_url = rabbitmq_secondary_url
+    def __init__(self, rabbitmq_url, rabbitmq_secondary_url, username, password):
+        self.rabbitmq_url = "amqp://{}:{}@{}/%2F".format(username, password, rabbitmq_url) if username and password \
```
Same as for the database - we're calling two different things `rabbitmq_url`

ajmaddaford commented on tests/app/authentication/test_authenticator.py:
```diff
@@ -17,7 +17,7 @@ def setUp(self):
         self.application = application
         # Use an in memory database
-        self.database = Database("sqlite://", 1, 0)
+        self.database = Database("sqlite", "", None, None, 1, 0)
```
We should use named args when instantiating Database as it's unclear what all the arguments are for. It's also worth considering moving the retry count and retry delay values into constants in database.py

ajmaddaford commented on app/__init__.py:
```diff
@@ -63,8 +72,11 @@ def create_app():  # noqa: C901  pylint: disable=too-complex
-            application.config['EQ_RABBITMQ_URL_SECONDARY'],
+            rabbitmq_host=application.config['EQ_RABBITMQ_HOST'],
+            rabbitmq_secondary_host=application.config['EQ_RABBITMQ_HOST_SECONDARY'],
+            rabbitmq_port=application.config['EQ_RABBITMQ_PORT'],
```
I think we should drop the `rabbitmq_` from these params

ajmaddaford commented on app/__init__.py:
```diff
@@ -77,9 +89,12 @@ def create_app():  # noqa: C901  pylint: disable=too-complex
+        database_driver=application.config['EQ_SERVER_SIDE_STORAGE_DATABASE_DRIVER'],
+        database_host=application.config['EQ_SERVER_SIDE_STORAGE_DATABASE_HOST'],
+        database_port=application.config['EQ_SERVER_SIDE_STORAGE_DATABASE_PORT'],
+        database_name=application.config['EQ_SERVER_SIDE_STORAGE_DATABASE_NAME'],
```
I think we should drop the `database_` from the driver, host and port params. `database_name` should stay though as it's less clear what that is if we drop the `database_` from it.

ajmaddaford commented on app/data_model/database.py:
```diff
@@ -65,21 +65,36 @@ def __init__(self, jti_claim):
-    def __init__(self, database_url, database_setup_retry_count, database_setup_retry_delay):
+    def __init__(self, database_driver, database_name, database_host=None, database_port=None,
+                 username=None, password=None, database_setup_attempts=DATABASE_SETUP_ATTEMPTS,
+                 database_setup_retry_delay=DATABASE_SETUP_RETRY_DELAY_SECONDS):
```
I think we should also drop `database_` from the attempts and delay

ajmaddaford commented on app/submitter/submitter.py:
```diff
@@ -16,9 +16,19 @@ def send_message(self, message, queue, tx_id):  # pylint: disable=unused-argumen
+                                                                                              host=rabbitmq_secondary_host,
+                                                                                              port=rabbitmq_port) \
+            if username and password else "amqp://{host}:{port}/%2F".format(host=rabbitmq_secondary_host,
+                                                                            port=rabbitmq_port)
```
I think this would read more easily if the username and password conditional was moved to the top level, something like:
```
if username and password:
    ...
else:
    ...
```

ajmaddaford commented on docker-compose.yml:
```diff
@@ -31,9 +31,10 @@ services:
+      EQ_SECRETS_FILE: ".dev_secrets.yml"
+      EQ_SERVER_SIDE_STORAGE_DATABASE_URL: digitaleq:digitaleq@db:5432/runner
+      EQ_RABBITMQ_URL: digitaleq:digitaleq@rabbit:5672
+      EQ_RABBITMQ_URL_SECONDARY: digitaleq:digitaleq@rabbit:5672
```
I think these need to be replaced with the new host and port environment variables

ajmaddaford commented on tests/app/data_model/test_database_setup.py:
```diff
@@ -47,11 +52,32 @@ def test_database_retry_success_after_attempts(self):
+
+            # pylint: disable=protected-access
+            # Then a value is successfully returned
+            self.assertEqual(database._database_url, 'postgresql://testUsername:testPassword@localhost:5432/digitaleqrds')
```
If the database url generation was factored out into it's own method, we wouldn't need to mock the database to test it.

ajmaddaford commented on tests/app/submitter/test_submitter.py:
```diff
@@ -9,10 +9,11 @@
-        self.url1 = 'amqp://localhost:5672/%2F'
-        self.url2 = 'amqp://localhost:5672/%2F'
+        self.url1 = 'localhost'
+        self.url2 = 'localhost'
```
These should be host1 and host2

jonnyshaw89 committed:
Load secrets from yml file
Validate required secrets
Generate secrets using script on run to mimic behaviour of production deploy process

ajmaddaford commented on scripts/generate_secrets.py:
```diff
@@ -0,0 +1,48 @@
+
+keys = {}
+
+keys['EQ_SUBMISSION_SDX_PUBLIC_KEY'] = get_file_contents(keys_folder + 'sdc-submission-encryption-sdx-public-key.pem')
```
Path concatenation should be done with `os.path.join`, this will provide more flexibility with what input will be allowed (currently the trailing slash is required). Also, it's probably worth doing the concatenation in `get_file_contents` as currently every method is passing keys_folder.

ajmaddaford commented on scripts/generate_secrets.py:
```diff
@@ -0,0 +1,48 @@
+    return data
+
+
+keys_folder = sys.argv[1]
```
It's probably worth printing an error and exiting if no arg is passed
